feat: EPI-1124: Ability to manually enter the price of an additional invoice product

huytranbes opened the pull request:
### Changes

Ability to manually enter the price of an additional invoice product

### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

huytranbes committed:
feat: EPI-1124: Ability to manually enter the price of an additional invoice product

tcaiger commented on packages/web/app/components/Invoice/EditInvoiceModal/InvoiceItem.jsx:
```diff
@@ -115,6 +117,7 @@ export const InvoiceItemRow = ({
       casing: 'lower',
     },
   );
+  const hidePriceInput = item.product?.price || item.product?.price === 0;
```
I wonder if this would be better to be `item.product?.price !== undefined`. I'll let you decide

huytranbes replied:
No. The data from back-end actually returns null

tcaiger commented on packages/web/app/components/Invoice/EditInvoiceModal/InvoiceItem.jsx:
```diff
@@ -256,9 +260,17 @@ export const InvoiceItemRow = ({
+      ...item,
+      productPrice: defaultItem?.productPrice,
+    });
+  }, []);
```
```suggestion
  }, [defaultItem?.productPrice]);
```
Should this reset everytime the default item changes?

tcaiger commented on packages/web/app/components/Invoice/EditInvoiceModal/InvoiceItem.jsx:
```diff
@@ -345,17 +357,34 @@ export const InvoiceItemRow = ({
+              </>
+            ) : (
+              item.productId && (
+                <Field
+                  name={`invoiceItems.${index}.productPrice`}
+                  component={NumberField}
+                  min={0}
+                  max={999999}
+                  onInput={validateDecimalPlaces}
+                  size="small"
+                  required
+                  style={{ width: '100%' }}
+                />
```
```suggestion

const PriceField = ({ field, ...props }) => {
  const handleChange = e => {
    let { value } = e.target;
    // If value is empty string and negative sign just return
    if (value === '' || /^[−-]/.test(value)) {
      field.onChange(value);
      return;
    }

    // IF the value is a number, set it to 2 decimal places
    let num = parseFloat(value);
    if (!isNaN(num)) {
      field.onChange(num.toFixed(2));
    }
  };

  return <NumberField value={field.value} onInput={handleChange} step="0.01" {...props} />;
};

                <Field
                  name={`invoiceItems.${index}.productPrice`}
                  component={PriceField}
                  min={0}
                  max={999999}
                  onInput={validateDecimalPlaces}
                  size="small"
                  required
                  style={{ width: '100%' }}
                />
```

It might be better to make a custom field time that extends Number field. Something like this (I haven't tested this code) because I think that setting e.target.value could have unexpected consequences.

huytranbes replied:
I've tried this, but `field.onChange` didn't directly prevent typing invalid number when input
I've implement the `PriceField` component btw

tcaiger reviewed (changes requested):
Nice work - I left a couple of suggestions. I think that a PriceField would make the intention of the input much clearer and could be useful in other places in the Invoicing section of Tamanu

huytranbes committed:
add useEffect dependency and implement PriceField component

tcaiger commented on packages/web/app/components/Field/PriceField.jsx:
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import { NumberInput } from './NumberField';
+
+export const PriceField = ({ field, ...props }) => {
```
Sorry I am not that experienced with Formik so I had to look up the exact syntax. It turns out you can either use `const { setFieldValue } = useFormikContext();` or onChange with the the event object `field.onChange({ target: { value: parseFloat(value).toFixed(2), name: field.name } });`. I think that setFieldValue is a bit cleaner eg.

```
export const PriceField = ({ field, ...props }) => {
  const { setFieldValue } = useFormikContext();

  const handleChange = e => {
    const value = e.target.value;

    // If value is negative just return empty
    if (/^[−-]/.test(value)) {
      setFieldValue(value);
      return;
    }
    // If the value is a decimal number, set it to 2 decimal places
    if (value.includes('.')) {
      const decimalPlaces = value.split('.')[1].length;
      if (decimalPlaces > 2) {
        setFieldValue(field.name, parseFloat(value).toFixed(2));
      }
    }
  };

  return (
    <NumberInput
      {...props}
      value={field.value}
      name={field.name}
      onChange={handleChange}
      step="0.01"
      min={0}
      max={999999}
      size="small"
    />
  );
};

huytranbes replied:
I'm sorry, this didn't work either. Seems like Formik's method can not directly interfere with the DOM to prevent invalid input

tcaiger commented:
Also it looks like there are some tests failing so you will need to run the `dbt-generate-docs` script to get the tests to pass.

huytranbes committed:
update dbt model docs

huytranbes committed:
feat: Update price field validation to make it optional

huytranbes committed:
feat: Update price field validation for InvoiceProduct

huytranbes committed:
remove actions column in Patient payments

huytranbes committed:
fix: Update visibility logic for price input in InvoiceItem component

This commit modifies the visibility logic for the price input field in the `InvoiceItemRow` component. The price input will now also be hidden when the item is not editable, enhancing the user experience by preventing unnecessary input fields from appearing in the UI.

huytranbes committed:
feat: Update price field validation and UI for InvoiceItem

The code changes in this commit update the price field validation and user interface for the `InvoiceItem` component. It adds support for optional price input and improves the display of crossed-out prices. The changes also include importing the `PriceField` component and the `REFERENCE_TYPES` constant from the `@tamanu/constants` module.

This commit addresses the following recent user commits:
- feat: Update price field validation for InvoiceProduct
- feat: Update price field validation to make it optional
- feat: EPI-1124: Ability to manually enter the price of an additional invoice product

It also takes into account the established commit message conventions in the repository.

huytranbes committed:
fix incorrect validation error message

huytranbes committed:
fix lint and display issue
